File: src/main/java/me/lokka30/levelledmobs/commands/subcommands/SpawnerSubCommand.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.commands.subcommands;

import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.UUID;
import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.util.MessageUtils;
import me.lokka30.levelledmobs.util.PaperUtils;
import me.lokka30.levelledmobs.util.SpigotUtils;
import me.lokka30.levelledmobs.util.Utils;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;

/**
 * Gives the user a specialized spawner that only spawns mobs within certain level criteria
 *
 * @author stumper66
 * @since 3.0.0
 */
public class SpawnerSubCommand extends SpawnerBaseClass implements Subcommand {

    public SpawnerSubCommand(final LevelledMobs main) {
        super(main);
        startingArgNum = 2;
    }

    final private List<String> allSpawnerOptions = Arrays.asList(
        "/name", "/customdropid", "/spawntype", "/giveplayer", "/lore", "/minlevel", "/maxlevel",
        "/delay",
        "/maxnearbyentities", "/minspawndelay", "/maxspawndelay", "/requiredplayerrange",
        "/spawncount", "/spawnrange", "/nolore"
    );

    @Override
    public void parseSubcommand(final LevelledMobs main, @NotNull final CommandSender sender,
        final String label, final String[] args) {
        commandSender = sender;
        messageLabel = label;

        if (!sender.hasPermission("levelledmobs.command.spawner")) {
            main.configUtils.sendNoPermissionMsg(sender);
            return;
        }

        if (args.length < 2) {
            showMessage("command.levelledmobs.spawner.usage");
            return;
        }

        OperationEnum operationEnum = switch (args[1].toLowerCase()) {
            case "copy" -> OperationEnum.COPY;
            case "info" -> OperationEnum.INFO;
            default -> OperationEnum.CREATE;
        };

        boolean hasGivePlayer = false;
        for (int i = 2; i < args.length; i++) {
            if ("/giveplayer".equalsIgnoreCase(args[i])) {
                hasGivePlayer = true;
                break;
            }
        }

        if ((!hasGivePlayer || operationEnum != OperationEnum.CREATE)
            && !(sender instanceof Player)) {
            final String messageName = operationEnum != OperationEnum.CREATE ?
                "common.players-only" : "command.levelledmobs.spawner.no-player";

            showMessage(messageName);
            return;
        }

        switch (args[1].toLowerCase()) {
            case "create" -> parseCreateCommand(args);
            case "copy" -> parseCopyCommand(args);
            case "info" -> parseInfoCommand(args);
        }
    }

    private void parseInfoCommand(final String @NotNull [] args) {
        final UUID playerId = ((Player) commandSender).getUniqueId();

        if (args.length == 2) {
            showMessage(main.companion.spawnerInfoIds.contains(playerId) ?
                "command.levelledmobs.spawner.info.status-enabled"
                : "command.levelledmobs.spawner.info.status-not-enabled");
            return;
        }

        if ("on".equalsIgnoreCase(args[2])) {
            if (main.companion.spawnerCopyIds.contains(playerId)) {
                // can't have both enabled.  We'll disable copy first
                copyGotDisabled(playerId);
            }

            main.companion.spawnerInfoIds.add(playerId);
            showMessage("command.levelledmobs.spawner.info.enabled");
        } else if ("off".equalsIgnoreCase(args[2])) {
            infoGotDisabled(playerId);
        }
    }

    private void parseCopyCommand(final String[] args) {
        if (!commandSender.hasPermission("levelledmobs.command.spawner.copy")) {
            main.configUtils.sendNoPermissionMsg(commandSender);
            return;
        }

        final UUID playerId = ((Player) commandSender).getUniqueId();

        if (args.length == 2) {
            showMessage(main.companion.spawnerCopyIds.contains(playerId) ?
                "command.levelledmobs.spawner.copy.status-enabled"
                : "command.levelledmobs.spawner.copy.status-not-enabled");
            return;
        }

        if ("on".equalsIgnoreCase(args[2])) {
            if (main.companion.spawnerInfoIds.contains(playerId)) {
                // can't have both enabled.  We'll disable info first
                infoGotDisabled(playerId);
            }

            main.companion.spawnerCopyIds.add(playerId);
            showMessage("command.levelledmobs.spawner.copy.enabled");
        } else if ("off".equalsIgnoreCase(args[2])) {
            copyGotDisabled(playerId);
        }
    }

    private void copyGotDisabled(final UUID playerId) {
        main.companion.spawnerCopyIds.remove(playerId);
        showMessage("command.levelledmobs.spawner.copy.disabled");
    }

    private void infoGotDisabled(final UUID playerId) {
        main.companion.spawnerInfoIds.remove(playerId);
        showMessage("command.levelledmobs.spawner.info.disabled");
    }

    private void parseCreateCommand(final String[] args) {
        hadInvalidArg = false;

        final CustomSpawnerInfo info = new CustomSpawnerInfo(main, messageLabel);
        if (commandSender instanceof Player) {
            info.player = (Player) commandSender;
        }

        // arguments with no values go here:
        for (int i = 1; i < args.length; i++) {
            final String arg = args[i];
            if ("/nolore".equalsIgnoreCase(arg)) {
                info.noLore = true;
                break;
            }
        }

        for (int i = 0; i < allSpawnerOptions.size() - 1; i++) {
            final boolean mustBeANumber = (i > 4);
            final String command = allSpawnerOptions.get(i);
            final String foundValue = getArgValue(command, args, mustBeANumber);
            if (hadInvalidArg) {
                return;
            }
            if (Utils.isNullOrEmpty(foundValue)) {
                continue;
            }

            switch (command) {
                case "/name" -> info.customName = foundValue;
                case "/customdropid" -> info.customDropId = foundValue;
                case "/lore" -> info.customLore = foundValue;
                case "/spawntype" -> {
                    try {
                        info.spawnType = EntityType.valueOf(foundValue.toUpperCase());
                    } catch (final Exception ignored) {
                        commandSender.sendMessage("Invalid spawn type: " + foundValue);
                        return;
                    }
                }
                case "/minlevel" -> info.minLevel = Integer.parseInt(foundValue);
                case "/maxlevel" -> info.maxLevel = Integer.parseInt(foundValue);
                case "/delay" -> info.delay = Integer.parseInt(foundValue);
                case "/maxnearbyentities" -> info.maxNearbyEntities = Integer.parseInt(foundValue);
                case "/minspawndelay" -> info.minSpawnDelay = Integer.parseInt(foundValue);
                case "/maxspawndelay" -> info.maxSpawnDelay = Integer.parseInt(foundValue);
                case "/requiredplayerrange" -> info.requiredPlayerRange = Integer.parseInt(foundValue);
                case "/spawncount" -> info.spawnCount = Integer.parseInt(foundValue);
                case "/spawnrange" -> info.spawnRange = Integer.parseInt(foundValue);
                case "/giveplayer" -> {
                    if (Utils.isNullOrEmpty(foundValue)) {
                        showMessage("command.levelledmobs.spawner.no-player-specified");
                        return;
                    }
                    try {
                        info.player = Bukkit.getPlayer(foundValue);
                    } catch (final Exception e) {
                        showMessage("common.player-offline", "%player%", foundValue);
                        return;
                    }
                    if (info.player == null) {
                        showMessage("common.player-offline", "%player%", foundValue);
                        return;
                    }
                }
            }
        }

        if (info.minLevel == -1 && info.maxLevel == -1) {
            showMessage("command.levelledmobs.spawner.no-level-specified");
            return;
        }

        if (info.player == null) {
            showMessage("command.levelledmobs.spawner.no-player-specified");
            return;
        }

        generateSpawner(info);
    }

    public void generateSpawner(final @NotNull CustomSpawnerInfo info) {
        if (info.customName != null) {
            info.customName = MessageUtils.colorizeAll(info.customName);
        }

        final ItemStack item = new ItemStack(Material.SPAWNER);
        final ItemMeta meta = item.getItemMeta();
        if (meta != null) {
            setMetaItems(meta, info, "LM Spawner");

            meta.getPersistentDataContainer()
                .set(info.main.namespacedKeys.keySpawner, PersistentDataType.INTEGER, 1);
            meta.getPersistentDataContainer()
                .set(info.main.namespacedKeys.keySpawnerMinLevel, PersistentDataType.INTEGER,
                    info.minLevel);
            meta.getPersistentDataContainer()
                .set(info.main.namespacedKeys.keySpawnerMaxLevel, PersistentDataType.INTEGER,
                    info.maxLevel);
            if (!Utils.isNullOrEmpty(info.customDropId)) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerCustomDropId,
                        PersistentDataType.STRING, info.customDropId);
            }
            if (info.spawnType != EntityType.UNKNOWN) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerSpawnType, PersistentDataType.STRING,
                        info.spawnType.toString());
            }
            if (info.spawnRange != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerSpawnRange,
                        PersistentDataType.INTEGER, info.spawnRange);
            }
            if (info.minSpawnDelay != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerMinSpawnDelay,
                        PersistentDataType.INTEGER, info.minSpawnDelay);
            }
            if (info.maxSpawnDelay != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerMaxSpawnDelay,
                        PersistentDataType.INTEGER, info.maxSpawnDelay);
            }
            if (info.spawnCount != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerSpawnCount,
                        PersistentDataType.INTEGER, info.spawnCount);
            }
            if (info.requiredPlayerRange != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerRequiredPlayerRange,
                        PersistentDataType.INTEGER, info.requiredPlayerRange);
            }
            if (info.maxNearbyEntities != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerMaxNearbyEntities,
                        PersistentDataType.INTEGER, info.maxNearbyEntities);
            }
            if (info.delay != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerDelay, PersistentDataType.INTEGER,
                        info.delay);
            }
            if (info.customName != null) {
                meta.getPersistentDataContainer()
                    .set(info.main.namespacedKeys.keySpawnerCustomName, PersistentDataType.STRING,
                        info.customName);
            }

            item.setItemMeta(meta);
        }

        int useInvSlotNum = info.player.getInventory().getHeldItemSlot();
        if (info.player.getInventory().getItem(useInvSlotNum) != null) {
            useInvSlotNum = -1;
        }

        if (useInvSlotNum == -1) {
            for (int i = 0; i <= 35; i++) {
                if (info.player.getInventory().getItem(i) == null) {
                    useInvSlotNum = i;
                    break;
                }
            }
        }

        if (useInvSlotNum == -1) {
            showMessage("command.levelledmobs.spawner.inventory-full", info.player);
            return;
        }

        info.player.getInventory().setItem(useInvSlotNum, item);
        final String playerName = main.getVerInfo().getIsRunningPaper() ?
            PaperUtils.getPlayerDisplayName(info.player)
            : SpigotUtils.getPlayerDisplayName(info.player);

        final List<String> message = getMessage(
            "command.levelledmobs.spawner.spawner-give-message-console",
            new String[]{"%minlevel%", "%maxlevel%", "%playername%"},
            new String[]{String.valueOf(info.minLevel), String.valueOf(info.maxLevel), playerName}
        );

        if (!message.isEmpty()) {
            Utils.logger.info(message.get(0).replace(main.configUtils.getPrefix() + " ", ""));
        }

        showMessage("command.levelledmobs.spawner.spawner-give-message", info.player);
    }

    @Override
    public List<String> parseTabCompletions(final LevelledMobs main,
        final @NotNull CommandSender sender, @NotNull final String[] args) {
        if (!sender.hasPermission("levelledmobs.command.spawner")) {
            return Collections.emptyList();
        }

        if (args.length <= 2) {
            return Arrays.asList("copy", "create", "info");
        }

        if ("create".equalsIgnoreCase(args[1])) {
            return tabCompletions_Create(args);
        } else if (("info".equalsIgnoreCase(args[1]) || "copy".equalsIgnoreCase(args[1]))
            && args.length == 3) {
            return Arrays.asList("on", "off");
        }

        return Collections.emptyList();
    }

    @NotNull private List<String> tabCompletions_Create(@NotNull final String @NotNull [] args) {
        if (!Utils.isNullOrEmpty(args[args.length - 2])) {
            switch (args[args.length - 2].toLowerCase()) {
                case "/spawntype" -> {
                    final List<String> entityNames = new LinkedList<>();
                    for (final EntityType entityType : EntityType.values()) {
                        entityNames.add(entityType.toString().toLowerCase());
                    }
                    return entityNames;
                }
                case "/delay" -> {
                    return Collections.singletonList("0");
                }
                case "/minspawndelay" -> {
                    return Collections.singletonList("200");
                }
                case "/maxspawndelay" -> {
                    return Collections.singletonList("800");
                }
                case "/maxnearbyentities", "/requiredplayerrange" -> {
                    return Collections.singletonList("16");
                }
                case "/spawncount", "/spawnrange" -> {
                    return Collections.singletonList("4");
                }
                case "/giveplayer" -> {
                    final List<String> players = new LinkedList<>();
                    for (final Player player : Bukkit.getOnlinePlayers()) {
                        players.add(player.getName());
                    }
                    players.sort(String.CASE_INSENSITIVE_ORDER);
                    return players;
                }
            }
        }

        return checkTabCompletion(allSpawnerOptions, args);
    }

    private enum OperationEnum {
        CREATE,
        COPY,
        INFO
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/commands/subcommands/SpawnerSubcommand.kt
package io.github.arcaneplugins.levelledmobs.commands.subcommands

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.tree.LiteralCommandNode
import io.papermc.paper.command.brigadier.CommandSourceStack
import java.util.UUID
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.MainCompanion
import io.github.arcaneplugins.levelledmobs.commands.MessagesHelper
import io.github.arcaneplugins.levelledmobs.util.MessageUtils.colorizeAll
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.command.CommandSender
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

/**
 * Gives the user a specialized spawner that only spawns mobs within certain level criteria
 *
 * @author stumper66
 * @since 3.0.0
 */
object SpawnerSubcommand : SpawnerBaseClass("levelledmobs.command.spawner") {
    override val description = "Various commands for creating spawner cubes."

    fun buildCommand() : LiteralCommandNode<CommandSourceStack> {
        return createLiteralCommand("spawner")
            .executes { ctx -> commandSender = ctx.source.sender
                showMessage("command.levelledmobs.spawner.usage")
                return@executes Command.SINGLE_SUCCESS
            }
            .then(createLiteralCommand("create")
                .then(createGreedyStringArgument("values")
                    .suggests { ctx, builder -> buildTabSuggestions(allSpawnerOptions, ctx, builder) }
                    .executes { ctx -> processResults(ctx, OperationEnum.CREATE)
                        return@executes Command.SINGLE_SUCCESS
                    })
            .executes { ctx -> processResults(ctx, OperationEnum.CREATE)
                return@executes Command.SINGLE_SUCCESS
            })
            .then(createLiteralCommand("copy")
                .then(createStringArgument("value")
                    .suggests { _, builder -> builder.suggest("on").suggest("off").buildFuture() }
                    .executes { ctx -> processResults(ctx, OperationEnum.COPY)
                        return@executes Command.SINGLE_SUCCESS
                    })
            .executes { ctx -> processResults(ctx, OperationEnum.COPY)
                return@executes Command.SINGLE_SUCCESS
            })
            .then(createLiteralCommand("info")
                .then(createStringArgument("value")
                    .suggests { _, builder -> builder.suggest("on").suggest("off").buildFuture() }
                    .executes { ctx -> processResults(ctx, OperationEnum.INFO)
                        return@executes Command.SINGLE_SUCCESS
                    })
            .executes { ctx -> processResults(ctx, OperationEnum.INFO)
                return@executes Command.SINGLE_SUCCESS
            })
            .build()
    }

    private fun processResults(
        ctx: CommandContext<CommandSourceStack>,
        operation: OperationEnum
    ) {
        commandSender = ctx.source.sender
        val sender = ctx.source.sender
        val args = splitStringWithQuotes(ctx.input, false)

        if (args.size <= 2) {
            when (operation){
                OperationEnum.CREATE -> {
                    showMessage("command.levelledmobs.spawner.usage")
                }
                OperationEnum.COPY -> {
                    val playerId = (sender as Player).uniqueId
                    showMessage(
                        if (MainCompanion.instance.spawnerCopyIds.contains(playerId))
                            "command.levelledmobs.spawner.copy.status-enabled"
                        else
                            "command.levelledmobs.spawner.copy.status-not-enabled"
                    )
                }
                OperationEnum.INFO -> {
                    val playerId = (sender as Player).uniqueId
                    showMessage(
                        if (MainCompanion.instance.spawnerInfoIds.contains(playerId))
                            "command.levelledmobs.spawner.info.status-enabled"
                        else
                            "command.levelledmobs.spawner.info.status-not-enabled"
                    )
                }
            }
            return
        }

        var hasGivePlayer = false
        for (i in 0 until args.size) {
            if ("/giveplayer".equals(args[i], ignoreCase = true)) {
                hasGivePlayer = true
                break
            }
        }

        if ((!hasGivePlayer || operation != OperationEnum.CREATE)
            && sender !is Player
        ) {
            val messageName =
                if (operation != OperationEnum.CREATE)
                    "common.players-only" else "command.levelledmobs.spawner.no-player"

            MessagesHelper.showMessage(sender, messageName)
            return
        }

        when (operation) {
            OperationEnum.CREATE -> parseCreateCommand(sender, args)
            OperationEnum.COPY -> parseCopyCommand(ctx)
            OperationEnum.INFO -> parseInfoCommand(ctx)
        }
    }

    private fun parseInfoCommand(
        ctx: CommandContext<CommandSourceStack>
    ) {
        val sender = ctx.source.sender
        commandSender = sender
        val playerId = (sender as Player).uniqueId
        val main = LevelledMobs.instance
        val value = getStringArgument(ctx, "value")

        if ("on".equals(value, ignoreCase = true)) {
            if (main.mainCompanion.spawnerCopyIds.contains(playerId)) {
                // can't have both enabled.  We'll disable copy first
                copyGotDisabled(playerId)
            }

            main.mainCompanion.spawnerInfoIds.add(playerId)
            MessagesHelper.showMessage(sender, "command.levelledmobs.spawner.info.enabled")
        }
        else if ("off".equals(value, ignoreCase = true))
            infoGotDisabled(playerId)
    }

    private val allSpawnerOptions = mutableListOf(
        "/name", "/customdropid", "/spawntype", "/giveplayer",
        "/lore", "/minlevel", "/maxlevel", "/delay",
        "/maxnearbyentities", "/minspawndelay", "/maxspawndelay",
        "/requiredplayerrange", "/spawncount", "/spawnrange", "/nolore"
    )

    private fun parseCopyCommand(
        ctx: CommandContext<CommandSourceStack>
    ) {
        val main = LevelledMobs.instance
        val sender = ctx.source.sender
        val playerId = (sender as Player).uniqueId
        val value = getStringArgument(ctx, "value")

        if ("on".equals(value, ignoreCase = true)) {
            if (main.mainCompanion.spawnerInfoIds.contains(playerId)) {
                // can't have both enabled.  We'll disable info first
                infoGotDisabled(playerId)
            }

            main.mainCompanion.spawnerCopyIds.add(playerId)
            showMessage("command.levelledmobs.spawner.copy.enabled")
        }
        else if ("off".equals(value, ignoreCase = true))
            copyGotDisabled(playerId)

    }

    private fun copyGotDisabled(playerId: UUID) {
        MainCompanion.instance.spawnerCopyIds.remove(playerId)
        showMessage("command.levelledmobs.spawner.copy.disabled")
    }

    private fun infoGotDisabled(playerId: UUID) {
        MainCompanion.instance.spawnerInfoIds.remove(playerId)
        showMessage("command.levelledmobs.spawner.info.disabled")
    }

    private fun parseCreateCommand(
        sender: CommandSender,
        args: MutableList<String>
    ) {
        hadInvalidArg = false

        val info = CustomSpawnerInfo(false)
        if (sender is Player) info.player = sender

        // arguments with no values go here:
        for (i in 1 until args.size) {
            val arg = args[i]
            if ("/nolore".equals(arg, ignoreCase = true)) {
                info.noLore = true
                break
            }
        }

        for (i in 0 until allSpawnerOptions.size - 1) {
            val mustBeANumber = (i > 4)
            val command = allSpawnerOptions[i]
            val foundValue = getArgValue(command, args, mustBeANumber)
            if (hadInvalidArg) return

            if (foundValue.isNullOrEmpty()) continue

            when (command) {
                "/name" -> info.customName = foundValue
                "/customdropid" -> info.customDropId = foundValue
                "/lore" -> info.customLore = foundValue
                "/spawntype" -> {
                    try {
                        info.spawnType = EntityType.valueOf(foundValue.uppercase())
                    } catch (_: Exception) {
                        sender.sendMessage("Invalid spawn type: $foundValue")
                        return
                    }
                }

                "/minlevel" -> info.minLevel = foundValue.toInt()
                "/maxlevel" -> info.maxLevel = foundValue.toInt()
                "/delay" -> info.delay = foundValue.toInt()
                "/maxnearbyentities" -> info.maxNearbyEntities = foundValue.toInt()
                "/minspawndelay" -> info.minSpawnDelay = foundValue.toInt()
                "/maxspawndelay" -> info.maxSpawnDelay = foundValue.toInt()
                "/requiredplayerrange" -> info.requiredPlayerRange = foundValue.toInt()
                "/spawncount" -> info.spawnCount = foundValue.toInt()
                "/spawnrange" -> info.spawnRange = foundValue.toInt()
                "/giveplayer" -> {
                    if (foundValue.isEmpty()) {
                        MessagesHelper.showMessage(sender, "command.levelledmobs.spawner.no-player-specified")
                        return
                    }
                    try {
                        info.player = Bukkit.getPlayer(foundValue)
                    } catch (_: Exception) {
                        MessagesHelper.showMessage(sender, "common.player-offline", "%player%", foundValue)
                        return
                    }
                    if (info.player == null) {
                        MessagesHelper.showMessage(sender, "common.player-offline", "%player%", foundValue)
                        return
                    }
                }
            }
        }

        if (info.minLevel == -1 && info.maxLevel == -1) {
            showMessage("command.levelledmobs.spawner.no-level-specified")
            return
        }

        if (info.player == null) {
            showMessage("command.levelledmobs.spawner.no-player-specified")
            return
        }

        generateSpawner(info)
    }

    fun generateSpawner(
        info: CustomSpawnerInfo
    ) {
        if (info.customName != null)
            info.customName = colorizeAll(info.customName)

        val item = ItemStack(Material.SPAWNER)
        val meta = item.itemMeta
        val defaultName = "LM Spawner"
        setMetadata(item, info, defaultName)
        setMetaItems(meta, info, defaultName)
        giveItemToPlayer(item, info)
    }

    private enum class OperationEnum {
        CREATE,
        COPY,
        INFO
    }
}